Fix some sizing bugs with multiline text controls.

mpcomplete opened the pull request:
1. Properly calculate text height. Previously we were applying
   constraints to the text height, which was incorrect when there was a min
   height. This resulted, for example, a caret that was way too tall.
2. Scrollable now caches container and content extents more often. This
   prevents rebuilds from clobbering the previously caculated sizes, fixing
   a bug where you could scroll beyond the end of text.

Fixes #6329

mpcomplete commented:
I can't seem to write a test that repros these bugs. I can reliably repro them manually, but they don't occur. in the test framework.

mpcomplete commented:
OK, I found the magic incantation to trigger the bug, and verified that this patch fixes it.

abarth commented on packages/flutter/lib/src/widgets/scrollable.dart:
```diff
@@ -513,6 +513,9 @@ class ScrollableState<T extends Scrollable> extends State<T> with SingleTickerPr
   /// the given offset instead.
   void didUpdateScrollBehavior(double newScrollOffset) {
+    _contentExtent = scrollBehavior.contentExtent;
+    _containerExtent = scrollBehavior.containerExtent;
```
I guess this makes sense.  Probably should have @Hixie review though given that his head is more into scrolling at the moment.

Hixie replied:
@HansMuller is more familiar with this specific logic.

HansMuller replied:
This is tricky. The variables you're updating were added in https://github.com/flutter/flutter/pull/5862

If you actually change the _contentExtent here then you'd need to setState(). But I believe that if you wrap the changes to _contentExtent and _containerExtent in setState() you'll trip over an assertion.

Hixie commented on packages/flutter/test/widget/input_test.dart:
```diff
@@ -633,7 +633,11 @@ void main() {
-    await tester.pump();
+    await tester.pump(const Duration(seconds: 2));
+    // Wait and drag again to trigger https://github.com/flutter/flutter/issues/6329
+    // (No idea why this is necessary, but the bug wouldn't repro without it.)
```
Just file a bug describing how to trigger something manually, and point to this code saying that for some reason the test framework needs this to make it trigger. So long as the issue is clear enough about how to reproduce the difference, eventually I'll get around to investigating it.

mpcomplete replied:
The tricky part here is that it won't happen anymore with this fix. Or, presumably, with your changes to Scrollable. The only way to see that this is broken is running this code against the framework before this fix lands.

Hixie replied:
Can one can put a print somewhere in the code (even today's code) that shows a difference between the test and manual testing?

mpcomplete committed:
Add a maxLines parameter for multiline Input.

If maxLines is 1, it's a single line Input that scrolls horizontally.
Otherwise, overflowed text wraps and scrolls vertically, taking up at
most `maxLines`.

Also fixed scrolling behavior so that the Input scrolls ensuring the
cursor is always visible.

Fixes https://github.com/flutter/flutter/issues/6271

mpcomplete commented:
Using maybeSetState. Ready for another look.

HansMuller commented:
LGTM
